grakn-client/src/main/java/ai/grakn/remote/GrpcClient.java

```diff
@@ -0,0 +1,141 @@
+import ai.grakn.grpc.TxGrpcCommunicator;
+import ai.grakn.grpc.TxGrpcCommunicator.Response;
+import ai.grakn.rpc.generated.GraknGrpc;
+import ai.grakn.rpc.generated.GraknOuterClass.TxResponse;
+import ai.grakn.util.CommonUtil;
+import com.google.common.base.Throwables;
+import com.google.common.collect.AbstractIterator;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+
+/**
+ * Communicates with a Grakn gRPC server, translating requests and responses to and from their gRPC representations.
+ *
+ * <p>
+ *     This class is a light abstraction layer over gRPC - it understands how the sequence of calls should execute and
+ *     how to translate gRPC objects into Java objects and back. However, any logic is kept in {@link RemoteGraknTx}.
+ * </p>
+ *
+ * @author Felix Chapman
+ */
+final class GrpcClient implements AutoCloseable {
+
+    private final TxGrpcCommunicator communicator;
+
+    private GrpcClient(TxGrpcCommunicator communicator) {
+        this.communicator = communicator;
+    }
+
+    public static GrpcClient create(GraknGrpc.GraknStub stub) {
+        TxGrpcCommunicator observer = TxGrpcCommunicator.create(stub);
+        return new GrpcClient(observer);
+    }
+
+    public void open(Keyspace keyspace, GraknTxType txType) {
+        communicator.send(GrpcUtil.openRequest(keyspace, txType));
+        responseOrThrow();
+    }
+
+    public Iterator<Object> execQuery(Query<?> query, @Nullable Boolean infer) {
+        communicator.send(GrpcUtil.execQueryRequest(query.toString(), infer));
+
+        return new AbstractIterator<Object>() {
+            private boolean firstElem = true;
+
+            @Override
+            protected Object computeNext() {
+                if (firstElem) {
+                    firstElem = false;
+                } else {
+                    communicator.send(GrpcUtil.nextRequest());
+                }
+
+                TxResponse response = responseOrThrow();
+
+                switch (response.getResponseCase()) {
+                    case QUERYRESULT:
+                        return GrpcUtil.getQueryResult(response.getQueryResult());
+                    case DONE:
+                        return endOfData();
+                    default:
+                    case RESPONSE_NOT_SET:
+                        throw CommonUtil.unreachableStatement("Unexpected " + response);
+                }
+            }
+        };
+    }
+
+    public void commit() {
+        communicator.send(GrpcUtil.commitRequest());
+        responseOrThrow();
+    }
+
+    @Override
+    public void close() {
+        communicator.close();
+    }
+
+    private TxResponse responseOrThrow() {
+        Response response;
+
+        try {
+            response = communicator.receive();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw Throwables.propagate(e);
```

Comment:
What does this do? Also: https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate . . . Relevant?

Reply:
Ooh... I didn't know that. So it looks like I should just say `throw new RuntimeException(e);`. I hate `InterruptedException`s 😞